### PR TITLE
KAFKA-15629: Support ResultOrder to TimestampedRangeQuery.

### PR DESCRIPTION
**Pull request by hanyuzheng7**
Support `ResultOrder` to `TimestampedRangeQuery`.
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java [hunk 1]
```diff
@@ -59,25 +58,25 @@ private RangeQuery(final Optional<K> lower, final Optional<K> upper, final boole
...
-        return new RangeQuery<>(Optional.ofNullable(lower), Optional.ofNullable(upper), true);
+        return new RangeQuery<>(Optional.ofNullable(lower), Optional.ofNullable(upper), ResultOrder.ANY);
     }
 
     /**
-     * Determines if the serialized byte[] of the keys in ascending order.
+     * Determines if the serialized byte[] of the keys in ascending or any order.
      * Order is based on the serialized byte[] of the keys, not the 'logical' key order.
-     * @return true if ascending, false otherwise.
+     * @return return the order of returned records based on the serialized byte[] of the keys(can be unordered, or in ascending or in descending order).
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java]: missing space

File: streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java [hunk 2]
```diff
@@ -59,25 +58,25 @@ private RangeQuery(final Optional<K> lower, final Optional<K> upper, final boole
...
-        return isKeyAscending;
+    public ResultOrder resultOrder() {
+        return order;
     }
 
     /**
      * Set the query to return the serialized byte[] of the keys in descending order.
      * Order is based on the serialized byte[] of the keys, not the 'logical' key order.
-     * @return a new RangeQuery instance with descending flag set.
+     * @return a new RangeQuery instance with DESCENDING set.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java]: old description was better

File: streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java [hunk 3]
```diff
@@ -59,25 +58,25 @@ private RangeQuery(final Optional<K> lower, final Optional<K> upper, final boole
...
      */
     public RangeQuery<K, V> withDescendingKeys() {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java]: We should also add `withAscendingKeys()` method

File: streams/src/main/java/org/apache/kafka/streams/query/TimestampedRangeQuery.java [hunk 4]
```diff
@@ -82,25 +81,25 @@ public static <K, V> TimestampedRangeQuery<K, V> withUpperBound(final K upper) {
...
     public static <K, V> TimestampedRangeQuery<K, V> withLowerBound(final K lower) {
-        return new TimestampedRangeQuery<>(Optional.of(lower), Optional.empty(), true);
+        return new TimestampedRangeQuery<>(Optional.of(lower), Optional.empty(), ResultOrder.ANY);
     }
 
     /**
      * Determines if the serialized byte[] of the keys in ascending order.
      * Order is based on the serialized byte[] of the keys, not the 'logical' key order.
-     * @return true if ascending, false otherwise.
+     * @return return the order of return records base on the serialized byte[] of the keys(can be unordered, or in ascending, or in descending order).
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/query/TimestampedRangeQuery.java]: missing space

File: streams/src/main/java/org/apache/kafka/streams/query/TimestampedRangeQuery.java [hunk 5]
```diff
@@ -82,25 +81,25 @@ public static <K, V> TimestampedRangeQuery<K, V> withUpperBound(final K upper) {
...
-        return isKeyAscending;
+    public ResultOrder resultOrder() {
+        return order;
     }
 
     /**
      * Set the query to return the serialized byte[] of the keys in descending order.
      * Order is based on the serialized byte[] of the keys, not the 'logical' key order.
-     * @return a new RangeQuery instance with descending flag set.
+     * @return a new RangeQuery instance with DESCENDING set.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/query/TimestampedRangeQuery.java]: original description is better

File: streams/src/main/java/org/apache/kafka/streams/query/TimestampedRangeQuery.java [hunk 6]
```diff
@@ -82,25 +81,25 @@ public static <K, V> TimestampedRangeQuery<K, V> withUpperBound(final K upper) {
...
      */
     public TimestampedRangeQuery<K, V> withDescendingKeys() {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/query/TimestampedRangeQuery.java:101]: we should add `withAscendingKeys()`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java [hunk 7]
```diff
@@ -254,12 +255,12 @@ private <R> QueryResult<R> runRangeQuery(final Query<R> query,
...
         final RangeQuery<K, V> typedQuery = (RangeQuery<K, V>) query;
         RangeQuery<Bytes, byte[]> rawRangeQuery;
-        final boolean isKeyAscending = typedQuery.isKeyAscending();
+        final ResultOrder order = typedQuery.resultOrder();
         rawRangeQuery = RangeQuery.withRange(
                 keyBytes(typedQuery.getLowerBound().orElse(null)),
                 keyBytes(typedQuery.getUpperBound().orElse(null))
         );
-        if (!isKeyAscending) {
+        if (order.equals(ResultOrder.DESCENDING)) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java]: We need to handle the ascending case, too.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java [hunk 8]
```diff
@@ -205,12 +206,12 @@ private <R> QueryResult<R> runTimestampedRangeQuery(final Query<R> query,
...
         final TimestampedRangeQuery<K, V> typedQuery = (TimestampedRangeQuery<K, V>) query;
         RangeQuery<Bytes, byte[]> rawRangeQuery;
-        final boolean isKeyAscending = typedQuery.isKeyAscending();
+        ResultOrder order = typedQuery.resultOrder();
         rawRangeQuery = RangeQuery.withRange(
                 keyBytes(typedQuery.lowerBound().orElse(null)),
                 keyBytes(typedQuery.upperBound().orElse(null))
         );
-        if (!isKeyAscending) {
+        if (order.equals(ResultOrder.DESCENDING)) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java:214]: We need to handle the ascending case, too

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java [hunk 9]
```diff
@@ -268,12 +269,12 @@ private  <R> QueryResult<R> runRangeQuery(final Query<R> query,
...
         final RangeQuery<K, V> typedQuery = (RangeQuery<K, V>) query;
         RangeQuery<Bytes, byte[]> rawRangeQuery;
-        final boolean isKeyAscending = typedQuery.isKeyAscending();
+        ResultOrder order = typedQuery.resultOrder();
         rawRangeQuery = RangeQuery.withRange(
                 keyBytes(typedQuery.getLowerBound().orElse(null)),
                 keyBytes(typedQuery.getUpperBound().orElse(null))
         );
-        if (!isKeyAscending) {
+        if (order.equals(ResultOrder.DESCENDING)) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java]: Missing ascending case

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java (hunk 7 above)
Comment (hanyuzheng7) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java]: Given that 'ANY' and 'ASCENDING' are treated the same, except for 'ResultOrder.DESCENDING', should a separate statement be used to handle the 'ASCENDING' case?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java (hunk 8 above)
Comment (hanyuzheng7) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java:214]: https://github.com/apache/kafka/pull/14907#discussion_r1416481443

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java (hunk 9 above)
Comment (hanyuzheng7) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java]: https://github.com/apache/kafka/pull/14907#discussion_r1416481443

File: streams/src/main/java/org/apache/kafka/streams/query/TimestampedRangeQuery.java (hunk 6 above)
Comment (hanyuzheng7) [reply streams/src/main/java/org/apache/kafka/streams/query/TimestampedRangeQuery.java:101]: ok

File: streams/src/main/java/org/apache/kafka/streams/query/TimestampedRangeQuery.java (hunk 5 above)
Comment (hanyuzheng7) [reply streams/src/main/java/org/apache/kafka/streams/query/TimestampedRangeQuery.java]: ok

File: streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java (hunk 2 above)
Comment (hanyuzheng7) [reply streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java]: ok

File: streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java (hunk 1 above)
Comment (hanyuzheng7) [reply streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java]: ok

File: streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java (hunk 3 above)
Comment (hanyuzheng7) [reply streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java]: ok

File: streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java (hunk 1 above)
Comment (hanyuzheng7) [reply streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java]: But I didn't find where miss the space

File: streams/src/main/java/org/apache/kafka/streams/query/TimestampedRangeQuery.java [hunk 10]
```diff
@@ -100,17 +99,25 @@ public boolean isKeyAscending() {
      * @return a new RangeQuery instance with descending flag set.
      */
     public TimestampedRangeQuery<K, V> withDescendingKeys() {
-        return new TimestampedRangeQuery<>(this.lower, this.upper, false);
+        return new TimestampedRangeQuery<>(this.lower, this.upper, ResultOrder.DESCENDING);
     }
 
+    /**
+     * Set the query to return the serialized byte[] of the keys in Ascending order.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/query/TimestampedRangeQuery.java]:
```suggestion
     * Set the query to return the serialized byte[] of the keys in ascending order.
```

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java (hunk 7 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java]: This is covered with the previous PR, right?

**Pull request merged**